### PR TITLE
set up GPU environment

### PR DESCRIPTION
**Pull request by facaiy**
First step of  #118 .
Refer to GPU op example in https://github.com/tensorflow/custom-op

### CODE REVIEW

File: configure.sh [hunk 1]
```diff
@@ -53,3 +53,19 @@ write_action_env_to_bazelrc "TF_HEADER_DIR" ${TF_CFLAGS:2}
...
+write_to_bazelrc "build --spawn_strategy=standalone"
+write_to_bazelrc "build --strategy=Genrule=standalone"
+write_to_bazelrc "build -c opt"
+write_action_env_to_bazelrc "TF_NEED_CUDA" ${TF_NEED_CUDA}
+# TODO(yifeif): do not hardcode path
+if [[ "$TF_NEED_CUDA" == "1" ]]; then
+  write_action_env_to_bazelrc "CUDNN_INSTALL_PATH" "/usr/lib/x86_64-linux-gnu"
+  write_action_env_to_bazelrc "TF_CUDA_VERSION" "10.0"
+  write_action_env_to_bazelrc "TF_CUDNN_VERSION" "7"
+  write_action_env_to_bazelrc "CUDA_TOOLKIT_PATH" "/usr/local/cuda"
```
Comment (seanpmorgan) [inline configure.sh]: How do we feel about using $CUDA_HOME here?

File: tensorflow_addons/custom_ops/image/BUILD [hunk 2]
```diff
@@ -35,11 +78,11 @@ cc_binary(
     copts = [
         "-pthread",
         "-std=c++11",
-        D_GLIBCXX_USE_CXX11_ABI,
-    ],
+        "-D_GLIBCXX_USE_CXX11_ABI=0",
```
Comment (seanpmorgan) [inline tensorflow_addons/custom_ops/image/BUILD]: Hmmm why remove the configurable ABI flag here?

File: tensorflow_addons/utils/test_utils.py [hunk 3]
```diff
@@ -26,6 +26,7 @@
 from tensorflow.python.framework.test_util import run_all_in_graph_and_eager_modes
 from tensorflow.python.framework.test_util import run_deprecated_v1
 from tensorflow.python.framework.test_util import run_in_graph_and_eager_modes
+from tensorflow.python.framework.test_util import use_gpu
```
Comment (seanpmorgan) [inline tensorflow_addons/utils/test_utils.py]: This seems fairly straightforward to just fork?
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/test_util.py#L1420

File: configure.sh (hunk 1 above)
Comment (seanpmorgan) [reply configure.sh]: Hmmm seems custom-op-gpu doesn't have this ENV variable though I do think it's pretty standard practice:
https://numba.pydata.org/numba-doc/dev/cuda/overview.html#setting-cuda-installation-path

**Comment by facaiy** [comment]
@seanpmorgan Yeah, Sean. The PR is a draft, and most of codes are just copied from the gpu example of custom-op. I want to use it to check whether our GPU environment is ready. If all tests pass, then I'll like to reconstruct all the codes and invite everyone to take a look :-)

File: configure.sh (hunk 1 above)
Comment (facaiy) [reply configure.sh]: Agree, I'll fix them later after we fix the environment issues :-)

**Comment by facaiy** [comment]
@av8ramit Amit, can you compare internal image with tensorflow/tensorflow:custom-op-gpu? Does they share the same CUDA setting?

> Cuda Configuration Error: No library found under: /usr/lib/x86_64-linux-gnu/lib64/libcudnn.so.7, /usr/lib/x86_64-linux-gnu/lib64/stubs/libcudnn.so.7, /usr/lib/x86_64-linux-gnu/lib/powerpc64le-linux-gnu/libcudnn.so.7, /usr/lib/x86_64-linux-gnu/lib/x86_64-linux-gnu/libcudnn.so.7, /usr/lib/x86_64-linux-gnu/lib/x64/libcudnn.so.7, /usr/lib/x86_64-linux-gnu/lib/libcudnn.so.7, /usr/lib/x86_64-linux-gnu/libcudnn.so.7

**Comment by facaiy** [comment]
cc @yifeif @gunan

**Comment by seanpmorgan** [comment]
Friendly bump as this is blocking one our of milestone issues.
Is cuDNN installed on the GCP image? If so we can set an [ENV variable in our GPU CI script](https://github.com/tensorflow/addons/blob/master/tools/ci_testing/addons_gpu.sh) 

cc @karmel

**Comment by yifeif** [comment]
I assume there are some minor differences between the setup, and we are also in the middle of moving everything to centos (cc: @av8ramit).
Does addon use tensorflow/tensorflow:custom-op-gpu for release, GCP image for build? Or GCP image for both?

**Comment by seanpmorgan** [comment]
> I assume there are some minor differences between the setup, and we are also in the middle of moving everything to centos (cc: @av8ramit).
> Does addon use tensorflow/tensorflow:custom-op-gpu for release, GCP image for build? Or GCP image for both?

So our current setup is to use travis and custom-op images for build/release and GCP for CI testing (travis does not have GPU servers).

Writing this out though we'll need to confirm we can compile with nvcc on a no GPU server (I believe it's possible but haven't tried.)

**Comment by av8ramit** [comment]
I'm currently in the process of moving everything to CentOS. As for the current Ubuntu images I'm not sure of any significant differences. As I understand it they are designed to be fairly similar to each other. Any commands or tests I can run on the GCP instances for you?

**Comment by facaiy** [comment]
I think we'd better use the same image for both build/release and CI testing. It seems that tensorflow/tensorflow:custom-op-gpu works well right now. As for GCP image, is it possible that we can get it from somewhere? Or, can we use tensorflow/tensorflow:custom-op-gpu for CI testing?

**Comment by facaiy** [comment]
I find it really difficult to debug if we cannot get access to the CI testing image.

**Comment by av8ramit** [comment]
Hey @facaiy yeah I agree. Unfortunately this is for security reasons as well as a route to provide easy debugging internally. I'm not opposed to having your entire setup run on  tensorflow/tensorflow:custom-op-gpu.

**Comment by gunan** [comment]
Unfortunately, I cannot give you our GCP images.
However, the custom op docker image has exactly the same toolchains set up.

As far as I can see, for addons we only have ubuntu/linux builds set up.
I am OK with changing the build setup for addons repository so every build runs inside the docker container.

**Comment by seanpmorgan** [comment]
I have no objections to using the custom-op docker images for travis and kokoro. ITbh that'd be ideal so we have a full understanding of our build env.

**Comment by facaiy** [comment]
@av8ramit @gunan Thanks for the input, Amit, Gunhan. We agree to switch to custom-op-gpu, so what do we need to do?

**Comment by facaiy** [comment]
Gently ping @av8ramit :-)

cc @karmel for visibility.

**Comment by av8ramit** [comment]
I'm not sure what you mean by your question. Do you mean what to adjust to use the docker container?

I guess the next step would be to make `addons_gpu.sh` invoke and run tests in the corresponding docker container. That's what is being invoked on our side internally.

We can also modify how it's being invoked internally. Right now it's being called in a virtualenv.

**Comment by facaiy** [comment]
Yeah, I want to know what to do if we'll use custom-op-gpu image for CI testing. 

> I guess the next step would be to make addons_gpu.sh invoke and run tests in the corresponding docker container.

Agree, let's run `addons_gpu.sh` in custom-op-gpu container. Can you set up it, please?

> Right now it's being called in a virtualenv.

Do you mean the script is invoked in a virtual environment inside a docker container? It sounds good, because it's convenient to support both python 2 and 3 when using custom-op image. cc @seanpmorgan any thoughts, Sean?

**Comment by gunan** [comment]
I think what Amit wanted to say was, kokoro directly calls `addons_gpu.sh`
You are free to make any changes to it, and the CI will just pick that up.
You can edit that script right here:
https://github.com/tensorflow/addons/blob/master/tools/ci_testing/addons_gpu.sh

**Comment by facaiy** [comment]
>  kokoro directly calls addons_gpu.sh

I agree, and I think it's a good idea. But how to run it in the corresponding docker image? I think we cannot do it without help from Amit. Or do you suggest to call `docker run` command in the addons_gpu.sh by ourselves? It looks unsafe to me.

**Comment by gunan** [comment]
Yes, i do recommend running docker run from addons_gpu. As long as it is
properly reviewed, i think it should be fine.

**Comment by facaiy** [comment]
I see, either way is fine. Thanks for your input, Gunhan.

@seanpmorgan @av8ramit @karmel Hi, Sean, Amit, Karmel, what do you think of them?

**Comment by seanpmorgan** [comment]
> I see, either way is fine. Thanks for your input, Gunhan.
> 
> @seanpmorgan @av8ramit @karmel Hi, Sean, Amit, Karmel, what do you think of them?

I have no issues calling the docker image. Though I would prefer we also call custom-op in our cpu testing script as well for consistency.

**Comment by seanpmorgan** [comment]
@facaiy tests are now passing in the GPU containers. We can start working on this again.

**Comment by facaiy** [comment]
> @facaiy tests are now passing in the GPU containers. We can start working on this again.

Good news, Sean, I'll start working on the PR, and tell you when ready.

File: tensorflow_addons/custom_ops/image/BUILD (hunk 2 above)
Comment (facaiy) [reply tensorflow_addons/custom_ops/image/BUILD]: restored

File: tensorflow_addons/utils/test_utils.py (hunk 3 above)
Comment (facaiy) [reply tensorflow_addons/utils/test_utils.py]: done :-)

File: configure.sh (hunk 1 above)
Comment (facaiy) [reply configure.sh]: I leave a TODO comment here, and want to solve the problem in the following PRs. What do you think?

**Comment by facaiy** [comment]
@seanpmorgan @yifeif Sean, Yifei, can you take a look? I prefer to solve  https://github.com/tensorflow/addons/issues/118 issue step by step, so I try to make the PR a minimum change (most of the code is from https://github.com/tensorflow/custom-op, thank Yifei) , and it just works.

cc @Squadrick who might be interested.

File: WORKSPACE [hunk 4]
```diff
@@ -1,5 +1,7 @@
 load("//tf_dependency:tf_configure.bzl", "tf_configure")
+load("//external/gpu:cuda_configure.bzl", "cuda_configure")
```
Comment (facaiy) [inline WORKSPACE]: @seanpmorgan Sean, perhaps we could move tf_dependency to external directory as well, any idea?

File: tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op.cc [hunk 5]
```diff
@@ -77,17 +77,18 @@ TF_CALL_double(REGISTER);
 
 #undef REGISTER
 
-#if GOOGLE_CUDA
+// TODO: fix compile issue of the gpu kernel.
+#if 0 && GOOGLE_CUDA
```
Comment (facaiy) [inline tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op.cc]: @mels630 Matt, seems that we can migrate gpu kernel after the PR is merged, right?

File: configure.sh [hunk 6]
```diff
@@ -53,3 +53,21 @@ write_action_env_to_bazelrc "TF_HEADER_DIR" ${TF_CFLAGS:2}
...
+write_to_bazelrc "build -c opt"
+write_action_env_to_bazelrc "TF_NEED_CUDA" ${TF_NEED_CUDA}
+
+# TODO(yifeif): do not hardcode path
+if [[ "$TF_NEED_CUDA" == "1" ]]; then
+    # TODO: use CUDA_HOME here?
+    write_action_env_to_bazelrc "CUDNN_INSTALL_PATH" "/usr/lib/x86_64-linux-gnu"
+    write_action_env_to_bazelrc "TF_CUDA_VERSION" "10.0"
+    write_action_env_to_bazelrc "TF_CUDNN_VERSION" "7"
+    write_action_env_to_bazelrc "CUDA_TOOLKIT_PATH" "/usr/local/cuda"
```
Comment (seanpmorgan) [inline configure.sh]: Could we make this an ENV variable with a default
`${CUDA_HOME:=/usr/local/cuda}`

File: configure.sh [hunk 7]
```diff
@@ -53,3 +53,21 @@ write_action_env_to_bazelrc "TF_HEADER_DIR" ${TF_CFLAGS:2}
 write_action_env_to_bazelrc "TF_SHARED_LIBRARY_DIR" ${SHARED_LIBRARY_DIR}
 write_action_env_to_bazelrc "TF_SHARED_LIBRARY_NAME" ${SHARED_LIBRARY_NAME}
 write_action_env_to_bazelrc "TF_CXX11_ABI_FLAG" ${TF_CXX11_ABI_FLAG}
+
+write_to_bazelrc "build:cuda --define=using_cuda=true --define=using_cuda_nvcc=true"
+write_to_bazelrc "build:cuda --crosstool_top=@local_config_cuda//crosstool:toolchain"
+write_to_bazelrc "build --spawn_strategy=standalone"
+write_to_bazelrc "build --strategy=Genrule=standalone"
+write_to_bazelrc "build -c opt"
```
Comment (seanpmorgan) [inline configure.sh]: Probably can remove this as we use it in our CLI. That or we can remove it from our documentation.

File: WORKSPACE (hunk 4 above)
Comment (seanpmorgan) [reply WORKSPACE]: I agree they should both live in a sub-directory. How would you feel about `build_deps`?

**Comment by seanpmorgan** [comment]
Oh could we also update `CONTRIBUTING.md` to explain when to use `custom-op-gpu` container

File: tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op.cc (hunk 5 above)
Comment (mels630) [reply tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op.cc]: no objection to fixing in a later PR. As long as we can avoid losing track of the issue!

File: WORKSPACE (hunk 4 above)
Comment (seanpmorgan) [reply WORKSPACE]: Would you prefer we move tf_dependency into `build_deps` in a separate PR?

**Comment by seanpmorgan** [comment]
So tests are failing OOM because we launch parallel jobs on a single card. Once #344 is merged we'll see that everything is working though 2 new tests fail when running on GPU:

https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/beam_search_ops_test.py#L73

And all of the `testSparseRepeatedIndices` in:
https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/weight_decay_optimizers_test.py

I'm okay with commenting these tests out and cutting new issues to address them.

**Comment by facaiy** [comment]
Thank you, Sean. I have disabled all those failures temporally, and filed the corresponding issues to track them.

File: tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op.cc (hunk 5 above)
Comment (facaiy) [reply tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op.cc]: sure, file https://github.com/tensorflow/addons/issues/349 issue

File: WORKSPACE (hunk 4 above)
Comment (facaiy) [reply WORKSPACE]: totally agree

**Pull request merged**